common/template-parser.ts

```diff
@@ -139,6 +139,18 @@ export async function parseTemplate(
   }
 
   const ast = parser.parse(template, {}) as PNode[]
+
+  // hack: when we continue, remove the post along with the last newline from tree
+  if (opts.continue && ast.length > 1) {
```

Comment:
In unvalidated prompts the `{{post}}` placeholder can technically be anywhere.

Reply:
That's a good point! I have reexamined my whole approach to this problem. Our goal here is not to manipulate `{{post}}` per se but to cut the prompt as close to the last response in history as possible to let the model continue the generation.

**New algorithm:**
1. Manipulate ast so that the history is the last node (if present)
2. When rendering complex nodes (conditionals, iterables) cut strictly after the message was rendered

That means that stuff like `ujb` and whatever was written in validated/unvalidated presets after history (like modifiers) will be erased and it can affect the generation results. However, I have tested with different prompt formats (albeit with sub 1 temperatures) and this approach beats the state of art every time.

Reply:
We might be able to avoid modifying the template-parser if we send the correct history/lines from the frontend?

Reply:
I don't think it's possible. It works without any modification to the template if `{{history}}` is used. Then we can just cut the tree and it will be enough. The complexity arises from the usage of `{{each msg}}` because it allows users to modify how each individual line in history is rendered.

So, the line we pass from the front may be perfectly fine (e.g. `Hello world`) but upon render it will turn into `Hello world\n\n`. There could be any symbols at the end, not just new lines and after it's rendered it is impossible to determine whether they came from the template or from the message. This is a major cause of interference when it comes to continuing the message.

The only way I found to ensure that the prompt ends on what was sent by the bot is to change how history props are edited in template parser. https://github.com/agnaistic/agnai/blob/ce433fc6941c76e0d48288ff5cdfbf56ba2329e5/common/template-parser.ts#L489

Reply:
I guess an alternative would be to have a completely different "continue" template that is triggered in this particular case?